### ISSUE TITLE
Cannot compile Leksah from latest source into a cabal sandbox

### ISSUE THREAD

**Issue by srb623** [leksah/leksah#457]
I am running Debian Linux, 64 bit, testing release (ie buster), kernel 4.13.8
and have been trying to compile the latest Leksah from source (cloned with the
--recursive option etc, as per the instructions), into a cabal sandbox, and have
not been able to due to the following error:

`
[10 of 68] Compiling IDE.Core.State   ( src/IDE/Core/State.hs, dist/dist-sandbox-864c1549/build/IDE/Core/State.o )

src/IDE/Core/State.hs:202:13-28: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (PaneName, b3), b4)’
    • In the pattern: (Just (pn, _), _)
      In a case alternative:
          (Just (pn, _), _) | pn == paneName pane -> return ()
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Just (pn, _), _) | pn == paneName pane -> return ()
           (mbActive, panes)
             -> do deactivatePaneWithout
                   triggerEventIDE (StatusbarChanged [...])
                   ....’
    |
202 |             (Just (pn,_), _) | pn == paneName pane -> return ()
    |             ^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:203:13-29: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (Text, b5), [PaneName])’
    • In the pattern: (mbActive, panes)
      In a case alternative:
          (mbActive, panes)
            -> do deactivatePaneWithout
                  triggerEventIDE
                    (StatusbarChanged [CompartmentPane (Just (PaneC pane))])
                  bringPaneToFront pane
                  ....
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Just (pn, _), _) | pn == paneName pane -> return ()
           (mbActive, panes)
             -> do deactivatePaneWithout
                   triggerEventIDE (StatusbarChanged [...])
                   ....’
    |
203 |             (mbActive, panes) -> do
    |             ^^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:209:31-64: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (PaneName, Connections), [PaneName])’
    • In the first argument of ‘setActivePane’, namely
        ‘(Just (paneName pane, conn), mru')’
      In a stmt of a 'do' block:
        setActivePane (Just (paneName pane, conn), mru')
      In the expression:
        do deactivatePaneWithout
           triggerEventIDE
             (StatusbarChanged [CompartmentPane (Just (PaneC pane))])
           bringPaneToFront pane
           let mru = filter (/= paneName pane) panes
               mru' = maybe mru ((: mru) . fst) mbActive
           ....
    |
209 |                 setActivePane (Just (paneName pane, conn), mru')
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:(234,38)-(235,101): error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(a1, [PaneName])’
    • In the ‘activePane’ field of a record
      In the ‘frameState’ field of a record
      In the expression:
        ide
          {frameState = (frameState ide)
                          {activePane = (fst (activePane (frameState ide)),
                                         filter
                                           (/= paneName pane) (snd (activePane (frameState ide))))}}
    |
234 |                         activePane = ( fst (activePane (frameState ide))
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

src/IDE/Core/State.hs:234:45-71: error:
    • Couldn't match expected type ‘(a1, b6)’
                  with actual type ‘Maybe (PaneName, Connections)’
    • In the first argument of ‘fst’, namely
        ‘(activePane (frameState ide))’
      In the expression: fst (activePane (frameState ide))
      In the ‘activePane’ field of a record
    |
234 |                         activePane = ( fst (activePane (frameState ide))
    |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:235:72-98: error:
    • Couldn't match expected type ‘(a2, [PaneName])’
                  with actual type ‘Maybe (PaneName, Connections)’
    • In the first argument of ‘snd’, namely
        ‘(activePane (frameState ide))’
      In the second argument of ‘filter’, namely
        ‘(snd (activePane (frameState ide)))’
      In the expression:
        filter (/= paneName pane) (snd (activePane (frameState ide)))
    |
235 |                                      , filter (/= paneName pane) (snd (activePane (frameState ide))))}}
    |                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:423:9-19: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe a6, b8)’
    • In the pattern: (Nothing, _)
      In a case alternative: (Nothing, _) -> return ()
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Nothing, _) -> return ()
           (Just (pn, _), _)
             -> do deactivatePaneWithout
                   triggerEventIDE
                     (RecordHistory (PaneSelected Nothing, PaneSelected (Just pn)))
                   ....’
    |
423 |         (Nothing,_)      -> return ()
    |         ^^^^^^^^^^^

src/IDE/Core/State.hs:424:9-24: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (Text, b9), b10)’
    • In the pattern: (Just (pn, _), _)
      In a case alternative:
          (Just (pn, _), _)
            -> do deactivatePaneWithout
                  triggerEventIDE
                    (RecordHistory (PaneSelected Nothing, PaneSelected (Just pn)))
                  triggerEventIDE (Sensitivity [(SensitivityEditor, False)])
                  ....
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Nothing, _) -> return ()
           (Just (pn, _), _)
             -> do deactivatePaneWithout
                   triggerEventIDE
                     (RecordHistory (PaneSelected Nothing, PaneSelected (Just pn)))
                   ....’
    |
424 |         (Just (pn, _),_) -> do
    |         ^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:435:9-31: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (a4, Connections), [a4])’
    • In the pattern: (Just (n, signals), mru)
      In a case alternative:
          (Just (n, signals), mru)
            -> do signalDisconnectAll signals
                  setActivePane (Nothing, n : mru)
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Just (n, signals), mru)
             -> do signalDisconnectAll signals
                   setActivePane (Nothing, n : mru)
           (Nothing, _) -> return ()’
    |
435 |         (Just (n,signals), mru) -> do
    |         ^^^^^^^^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:437:27-42: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe a3, [a4])’
    • In the first argument of ‘setActivePane’, namely
        ‘(Nothing, n : mru)’
      In a stmt of a 'do' block: setActivePane (Nothing, n : mru)
      In the expression:
        do signalDisconnectAll signals
           setActivePane (Nothing, n : mru)
    • Relevant bindings include
        mru :: [a4] (bound at src/IDE/Core/State.hs:435:28)
        n :: a4 (bound at src/IDE/Core/State.hs:435:16)
    |
437 |             setActivePane (Nothing, n:mru)
    |                           ^^^^^^^^^^^^^^^^

src/IDE/Core/State.hs:438:9-20: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe a5, b7)’
    • In the pattern: (Nothing, _)
      In a case alternative: (Nothing, _) -> return ()
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Just (n, signals), mru)
             -> do signalDisconnectAll signals
                   setActivePane (Nothing, n : mru)
           (Nothing, _) -> return ()’
    |
438 |         (Nothing, _) -> return ()
    |         ^^^^^^^^^^^^

src/IDE/Core/State.hs:443:9-20: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe a0, b0)’
    • In the pattern: (Nothing, _)
      In a case alternative: (Nothing, _) -> return ()
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Nothing, _) -> return ()
           (Just (n, _), _) -> when (n == paneName pane) deactivatePane’
    |
443 |         (Nothing, _) -> return ()
    |         ^^^^^^^^^^^^

src/IDE/Core/State.hs:444:9-23: error:
    • Couldn't match expected type ‘Maybe (PaneName, Connections)’
                  with actual type ‘(Maybe (PaneName, b1), b2)’
    • In the pattern: (Just (n, _), _)
      In a case alternative:
          (Just (n, _), _) -> when (n == paneName pane) deactivatePane
      In the second argument of ‘(>>=)’, namely
        ‘\case
           (Nothing, _) -> return ()
           (Just (n, _), _) -> when (n == paneName pane) deactivatePane’
    |
444 |         (Just (n,_), _) -> when (n == paneName pane) deactivatePane
    |         ^^^^^^^^^^^^^^^
Failed to install leksah-0.16.2.2
cabal: Error: some packages failed to install:
leksah-0.16.2.2-IOxJfdagq91GCahVfKiXqh failed during the building phase. The
exception was:
ExitFailure 1
`

I am running the following version of ghc -
    The Glorious Glasgow Haskell Compilation System, version 8.2.1

and the following version of cabal -
    compiled using version 2.0.0.2 of the Cabal library.

**Comment by hamishmack** [comment]
Have you tried using the `leksah.sh` script as descried in the [Getting Leksah](https://github.com/leksah/leksah#getting-leksah) section of the readme file?

**Comment by srb623** [comment]
Yes, I have tried that.  That was how I used to build Leksah until I decided to sandbox it.  Then I was able to successfully build into a sandbox until I received the set of errors that I have copied above.  Not that I can really comment, given my very limited Haskell knowledge, but are you sure that there is not an error in the code ?  Its just that the messages look like compiler errors to me.

**Comment by hamishmack** [comment]
This error looks like what would happen it was using slightly old version of `ltk`.  I can't really tell why that would be the case without knowing what the `cabal` command was used to build.

The two most likely causes are:

* The `cabal` command used did not reference the `ltk` submodule.
* The `ltk` submodule was not up to date (you could run `git status` to check).  

You should not need to use `cabal sandbox` since the `leksah.sh` script uses `cabal new-build` (which has all the advantages of a cabal sandbox while still allowing sharing).

Please run `git submodule update` to make sure you have the correct `ltk` submodule and try following the instructions [Getting Leksah](https://github.com/leksah/leksah#getting-leksah) again.

**Comment by srb623** [comment]
Did a git submdoule update and tried to build using leksah.sh as per the instructions and the build failed again.  Between doing this and getting a chance to comment here I noted that you had made a commit re constraints on leksah and ltk and I noted from the above that you mentioned that my version of ltk may be the issue.  So I did another fetch from git and ran the build again as per the instructions and with the latest changes the build now runs to completion.  Closing this as the issue is now resolved.  Many thanks for your help.
